lib/improver/cli.py

```diff
@@ -0,0 +1,340 @@
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""IMPROVER CLI base class."""
+
+import sys
+from sets import Set
+import inspect
+
+from improver.argparser import ArgParser
+from improver.utilities.load import load_cube  # as DEFAULT_LOADER
+from improver.utilities.save import save_netcdf  # as DEFAULT_SAVER
+
+
+DEFAULT_LOADER = load_cube
+DEFAULT_SAVER = save_netcdf
+
+# argument_mapping looks like:
+# {'cli_argument': '__init__.argument'} etc..
+# {'input_filepath': 'LOADER.cube'}
+class CLI(object):
+
+    # TODO: this might be the order that the methods are called...?
+    SUPPORTED_PLUGIN_METHODS = ('__init__', 'process')
+
+    SPECIAL_FUNCTIONS = ('LOADER', 'SAVER')
+
+    def __init__(self, plugin, argument_mapping, cli_argspecs=None,
+                 description=None, loader=DEFAULT_LOADER, saver=DEFAULT_SAVER):
+        """Create a CLI for a plugin.
+        The ArgParser instance is created automatically."""
+
+        self.plugin = plugin
+        self.loader = loader
+        self.saver = saver
+
+        # We assume that all plugins are currently implemented as classes
+        # with __init__ and process methods.
+        # To enforce: check that the plugin has only methods which are
+        # contained within CLI.SUPPORTED_PLUGIN_METHODS - this will not be
+        # necessary once all plugins inherit from a common base class
+        self._check_plugin_conforms()
+
+        # for user friendliness, we choose the input argument_mapping to look
+        # like: { 'cli_argument': 'plugin_method.method_argument' }
+        # but processing is easier if our dictionary looks like:
+        # { 'plugin_method': { 'method_argument': 'cli_argument' } }
+        # ...let's restructure:
+        cli_method_args = self._reorder_mapping_dict(argument_mapping)
+
+        # There are some special functions which are treated differently
+        special_args = {}
```

Comment:
special_args = dict(special_function = cli_method_args.pop(special_function) 
                                for special_function in SPECIAL_FUNCTIONS)